Add error counters collection and metrics for vdev from ZFS

vk-en opened the pull request:
This PR adds:
- Collection of I/O metrics for zpool and physical disks included in the zpool (for all types of operations) from ZFS;
- Collection of read/write/checksum error counters that we can see from the output of the **_zpool status_** command;
- Collection of a more understandable state for zpool and devices(/dev/...), indicating the obvious problem that occurred, if any. This can also usually be observed from the output of the **_zpool status_** command (in a status string) if there are problems.

Signed-off-by: Vitaliy Kuznetsov <ahil52_25@mail.ru>

eriknordmark commented on pkg/pillar/types/zfs.go:
```diff
@@ -99,10 +169,37 @@ type DiskDescription struct {
+	ZIOTypeWrite
+	ZIOTypeFree
+	ZIOTypeClaim
+	ZIOTypeIOCtl
```
Why such off capitalization of ioctl?

eriknordmark commented on pkg/pillar/types/zfs.go:
```diff
@@ -44,6 +44,74 @@ func (status ZVolStatus) Key() string {
+
+// PoolStatus value
+const (
+	PoolStatusCorruptCache      PoolStatus = iota // PoolStatusCorruptCache - corrupt /kernel/drv/zpool.cache
```
Normally we reserve the zero enum value as a *Unused so handle 1) the case when it isn't set and 2) have 1-1 mappings from golang enum values to protobuf enum values where the convention is that zero is unused.

vk-en replied:
These values come from ZFS. 
Now I fixed it by adding Unspecified values and shifted by one.

eriknordmark replied:
OK, but please add a comment for this PoolStatus definition that the values come from XXX (some zfs header file?) but we add one to their values to that we can have a Unused=0 value to follow our conventions.
Basically that the definitions should follow that zfs header file.

vk-en replied:
Сorrected

eriknordmark commented on pkg/pillar/types/zfs.go:
```diff
@@ -85,6 +153,8 @@ type ZFSPoolStatus struct {
 	Disks            []*StorageDiskState
 	CollectorErrors  string
 	Children         []*StorageChildren
+	Metrics          *ZDeviceMetrics // metrics and error counters for zpool
+	PoolStatusMsg    PoolStatus      // pool status value from ZFS
```
Are we going to report the metrics to the controller using the metrics API i.e., periodic reports? I think we report the ZFSPoolStatus content using the event-triggered info API. Depending on the answer it might make sense to have a separate ZFSPoolMetrics struct which will contain the things which will be reported periodically.

vk-en replied:
Yes, it is in PR: #2648 
ZFSPoolStatus serves as a container in which we collect Zpool data. And in the information API there is no subscription to events its updates (Create, modify, delete). There, when calling, we get all available ZFSPoolStatus, collect an informational message from the data, and that's it. Therefore, I think that there is no need to move Metrics from ZFSPoolStatus to a separate ZFSPoolMetrics. But if this is still critical for us, let me know about it, and I change it.

Another question is how often we want to collect these metrics. After @pram answers this question, I can add a new go-routine at the right interval to update this information.

eriknordmark replied:
The philosophy is that Status changes due to some event (which could be either some configuration change, or some internally generated event such as an error), and as such are reported through the system (publish towards zedagent and then zedagent sending an Info message to the controller when such an event/change takes place.

Metrics are always periodic; collected and sent periodically.

The fact that you say how often we want to collect this tells me is should be a separate metrics struct.

vk-en replied:
Сorrected

eriknordmark commented on pkg/pillar/types/zfs.go:
```diff
@@ -99,10 +171,41 @@ type DiskDescription struct {
 	DiskName *DiskDescription
 	Status   StorageStatus
+	AuxState VDevAux
+	Metrics  *ZDeviceMetrics
```
See comment elsewhere about separating periodicly reported metrics and state changes.

vk-en replied:
Сorrected

vk-en commented on pkg/pillar/types/zfs.go:
```diff
@@ -99,10 +171,41 @@ type DiskDescription struct {
+	ZIOTypeFree
+	ZIOTypeClaim
+	ZIOTypeIoctl
+	ZIOTypes
```
Сorrected

eriknordmark commented on pkg/pillar/types/zfs.go:
```diff
@@ -83,6 +167,26 @@ const (
+// ZFSMetrics - metrics for one pool or dataset
+type ZFSMetrics struct {
+	DisplayName     string          // Pool or Dataset name
+	CollectionTime  int64           // Time in seconds when the metrics was collected
```
Better to use time.Time for this.

vk-en replied:
Сorrected

eriknordmark commented on pkg/pillar/types/zfs.go:
```diff
@@ -12,6 +13,9 @@ const (
 	// ZVolDevicePrefix controlled by mdev
 	ZVolDevicePrefix = "/dev/zvol"
 
+	// ZFSMetricsHistoryMax - max number of stored metrics
+	ZFSMetricsHistoryMax = 60
+
```
Why do we need this for zfs metrics? We do not have it for any other metrics (and 99.9% of our metrics are cumulative since boot hence we can calculate averges even if the controller is missing some metrics message.)

The reason this is important is since the metrics message has all of the metrics (for device, app instances, etc) and it is limited to about 32 to 64kbyte is size.

eriknordmark replied:
> I added this to store the history of metrics in Eve's memory, now will store the last 60 values, given that we do data collection every second, it turns out that in Eve there will always be access to metrics for the last minute.

Why do you need to collect them every second in the first place?
Aren't all/most of them cumulative counters since boot?

We haven't had any customer ask us about more frequent data but we do have customers asking how they can reduce the amount of bandwidth used (which we can do by having zedagent report metrics a lot less frequently than once a minute.)

vk-en replied:
> Why do you need to collect them every second in the first place?
> Aren't all/most of them cumulative counters since boot?

I have only one reason for this - the accuracy of the information displayed on the controller. The shorter the interval between the collection of information from the meters, the more real and accurate the metrics and performance of the storage system are displayed. And this makes it more likely to detect anomalies with hardware or software problems in the storage system.
Sending a slice of metrics with an interval of more than one minute has little useful meaning, except to report that the counters are increasing and the disk has not yet failed.

> We haven't had any customer ask us about more frequent data but we do have customers asking how they can reduce the amount of bandwidth used (which we can do by having zedagent report metrics a lot less frequently than once a minute.)

As for reducing the amount of bandwidth used, I think that here it will not be possible to sit on two chairs at the same time, observing the quality of the displayed metrics and the rarity of the frequency of sending them to the controller.
In this case, for clients who need to reduce the amount of bandwidth used, I can, depending on the value of the timer (by which we send metrics), do the following:
- Send metric slices in the full amount necessary for accurate calculations, if there is no need to reduce the amount of bandwidth used.
- If it is necessary to reduce the amount of bandwidth used, then it is possible to send a slice with a collection difference per minute or even less often (which is not of practical utility), or not to send this information for this configurations (rare messages with metric), since it is not critical information for our system.

eriknordmark replied:
> Sending a slice of metrics with an interval of more than one minute has little useful meaning, except to report that the counters are increasing and the disk has not yet failed.

Plus it also allows the controller/UI to determine the rate of change to the counters (by comparing two sets whether those sets are 1 minute or 1 hour apart).
This is how we do all the other metrics.

I see no reason why these ZFS counters need any different treatment than the other metrics.

vk-en replied:
Well, if that's by design then ok. I removed this value, and also simplified the code so as not to store anything, since we do not need to collect metrics with a difference per second.

eriknordmark commented on pkg/pillar/cmd/zfsmanager/zfsstoragemetrics.go:
```diff
@@ -0,0 +1,97 @@
+			}
+
+			pub, _ := ctxPtr.storageMetricsPub.Get(poolName)
+			if pub == nil {
```
If it was already published don't you need to re-publish it with the current values?

eriknordmark reviewed (approved):
Some nits to address but let's kick of tests.

eriknordmark commented:
Note that the yetus complaints can be ignored; they are from
"Unchanged files with check annotations Beta"

vk-en commented on pkg/pillar/cmd/zfsmanager/zfsstoragemetrics.go:
```diff
@@ -0,0 +1,97 @@
+			pub, _ := ctxPtr.storageMetricsPub.Get(poolName)
+			if pub == nil {
+				// create new pool in metrics
+				pooMetrics := new(types.ZFSPoolMetrics)
```
Corrected. Not relevant for the new version.

eriknordmark reviewed (approved):
LGTM

eriknordmark commented:
Looks like the unit tests are failing due to go.mod issues:
=== Errors
package github.com/lf-edge/eve/pkg/pillar/cas imports github.com/moby/sys/mountinfo from implicitly required module; to add missing requirements, run:
	go get github.com/moby/sys/mountinfo@v0.6.0

eriknordmark commented:
Sorry I just realized @giggsoff introduced the unit test failiure in https://github.com/lf-edge/eve/pull/2718. You seem to have a fix for it already since the test passed now.

eriknordmark reviewed (approved):
Run eden again

giggsoff commented:
Seems I come a bit late, but just to make a note:
in ZFSPoolMetrics I can see `"DisplayName": "/dev/sda"`, in ZFSPoolStatus I can see `"DiskName":{"Name":"/dev/sda9","LogicalName":"","Serial":"QM00001"}`. Perhaps it would be nice to sync the output.

vk-en commented:
> Seems I come a bit late, but just to make a note:
> in ZFSPoolMetrics I can see `"DisplayName": "/dev/sda"`, in ZFSPoolStatus I can see `"DiskName":{"Name":"/dev/sda9","LogicalName":"","Serial":"QM00001"}`. Perhaps it would be nice to sync the output.

Thanks. I corrected it.
I have added an additional condition for work with the root device in PR #2648.
